File: user-srv/src/Pi.User/Pi.User.Domain/AggregatesModel/TradingAccountAggregate/TradingAccount.cs
using Pi.Common.SeedWork;

namespace Pi.User.Domain.AggregatesModel.TradingAccountAggregate;
public class TradingAccount
{
    public TradingAccount(
        string customerCode,
        string tradingAccountNo,
        string accountTypeCode,
        string exchangeMarketId,
        string? marketingId,
        string accountStatus)
    {
        CustomerCode = customerCode.Trim();
        TradingAccountNo = tradingAccountNo.Trim();
        AccountTypeCode = accountTypeCode;
        ExchangeMarketId = exchangeMarketId;
        AccountStatus = accountStatus;
        MarketingId = marketingId?.Trim();
        CreditLineCurrency = "THB";
    }

    public string CustomerCode { get; private set; }
    public string TradingAccountNo { get; private set; }
    public string AccountTypeCode { get; private set; }
    public string ExchangeMarketId { get; private set; }
    public string? AccountStatus { get; private set; }
    public decimal CreditLine { get; private set; }
    public string CreditLineCurrency { get; private set; }
    public DateTime? CreditLineEffectiveDate { get; private set; }
    public DateTime? CreditLineEndDate { get; private set; }
    public string? MarketingId { get; private set; }
    public DateTime? AccountOpeningDate { get; private set; }
    public TradingAccount WithCreditLine(decimal creditLine, string currency, DateTime? effectiveDate, DateTime? endDate)
    {
        CreditLine = creditLine;
        CreditLineCurrency = currency;
        CreditLineEffectiveDate = effectiveDate;
        CreditLineEndDate = endDate;
        return this;
    }
}

File: user-srv-v2/internal/dto/trading_account.go
package dto

import (
	domain "github.com/pi-financial/user-srv-v2/internal/domain"
	"github.com/pi-financial/user-srv-v2/internal/utils"
)

type DepositWithdrawTradingAccountResponse struct {
	TradingAccountId string `json:"tradingAccountId"`
	CustomerCode     string `json:"customerCode"`
	TradingAccountNo string `json:"tradingAccountNo"`
	ProductName      string `json:"productName"`
}

type TradingAccount struct {
	TradingAccountNo   string                    `json:"tradingAccountNo"`
	AccountTypeCode    string                    `json:"accountTypeCode"`
	AccountType        string                    `json:"accountType"`
	ExchangeMarketId   string                    `json:"exchangeMarketId"`
	CreditLine         float64                   `json:"creditLine"`
	CreditLineCurrency string                    `json:"creditLineCurrency"`
	EffectiveDate      utils.DateOnly            `json:"effectiveDate"` // 2025-12-31
	EndDate            utils.DateOnly            `json:"endDate"`       // 2025-12-31
	MarketingId        string                    `json:"marketingId"`
	SaleLicense        string                    `json:"saleLicense"`
	OpenDate           utils.DateOnly            `json:"openDate"` // 2025-12-31
	AccountStatus      domain.TradeAccountStatus `json:"accountStatus" validate:"oneof=N C"`
	FrontName          string                    `json:"frontName"`
	EnableBuy          string                    `json:"enableBuy"`
	EnableSell         string                    `json:"enableSell"`
	EnableDeposit      string                    `json:"enableDeposit"`
	EnableWithdraw     string                    `json:"enableWithdraw"`
}

type (
	CreateTradingAccountRequest struct {
		TradingAccount
	}

	// CreateTradingAccountRequestValidationWrapper Helps to validate on all objects in the lists in the wrapper
	CreateTradingAccountRequestValidationWrapper struct {
		TradingAccount []CreateTradingAccountRequest `validate:"required,dive"`
	}
)

type GetTradingAccountByUserIdParams struct {
	Status string `query:"status" validate:"omitempty,oneof=N C"`
}
